ec2/utility.py

```diff
@@ -0,0 +1,134 @@
+########
+# Copyright (c) 2014 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+# built-in package
+import time
+
+# other packages
+from boto.ec2 import EC2Connection as EC2
+from boto.exception import EC2ResponseError, BotoServerError
+
+# ctx packages
+from cloudify import ctx
+from cloudify.exceptions import NonRecoverableError
+from cloudify.decorators import operation
+
+
+@operation
+def validate_state(instance, state, timeout_length, check_interval, **kwargs):
+    """ Check if an EC2 instance is in a particular state.
+
+    :param instance: And EC2 instance.
+    :param state: The state code (pending = 0, running = 16,
+                  shutting down = 32, terminated = 48, stopping = 64,
+                  stopped = 80
+    :param timeout_length: How long to wait for a positive answer
+           before we stop checking.
+    :param check_interval: How long to wait between checks.
+    :return: bool (True the desired state was reached, False, it was not.)
+    """
+
+    ctx.logger.debug("""(Node: {0}): Attempting state validation:
+                        instance id: {0}, state: {1}, timeout length: {2},
+                        check interval: {3}.""".format(instance.id, state,
+                                                       timeout_length,
+                                                       check_interval))
+
+    timeout = time.time() + timeout_length
+
+    while True:
+        if state == get_instance_state(instance):
+            ctx.logger.info("""(Node: {0}):
+                               Instance state validated: instance {0}."""
+                            .format(instance.state))
+            return True
+        elif time.time() > timeout:
+            ctx.logger.error("""(Node: {0}): Timed out during instance
```

Comment:
don't think you should be logging and raising the same exceptions. I think that raising a nonrecoverable error also logs it. please verify with Idan and ask for the best practice here.